pkg/webui/components/key-value-map/index.js

```diff
@@ -0,0 +1,106 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import React from 'react'
+import bind from 'autobind-decorator'
+import { defineMessages } from 'react-intl'
+
+import PropTypes from '../../lib/prop-types'
+import Button from '../button'
+import Entry from './entry'
+
+const m = defineMessages({
+  addEntry: 'Add entry',
+})
+
+@bind
+class KeyValueMap extends React.PureComponent {
+
+  handleEntryChange (index, newValues) {
+    const { onChange, value } = this.props
+    onChange(value.map((kv, i) => (
+      index !== i ? kv : { ...kv, ...newValues }
+    )))
+  }
+
+  removeEntry (index) {
+    const { onChange, onBlur, value } = this.props
+    onChange(value.filter((_, i) => (i !== index)) || [])
+
+    // Trigger also the onBlur event to enforce revalidation
+    onBlur()
```

Comment:
seems more like a hack to me, `<KeyValueMap />` knows nothing about validation.
also, `onBlur` is not in the prop-types

Reply:
Any idea to achieve this cleaner?